File: JavaScript/src/name-util.js
const defaultdict = require('defaultdict2');
const under = require('underscore');

function averageLettersImperative(names)
{
	let total = 0;
	for(const name of names)
	{
    	total += name.length;
	}
	return total / names.length || 0;
}

function averageLettersFunctional(names)
{                       
	return names.map(name => name.length)
	  		.reduce((total, length) => total + length, 0)
	    	/ names.length || 0;
}

function startLetterCountImperative(names)
{
	let letterCounts = defaultdict(0);
	
	for(const name of names)
	{
		letterCounts[name[0]]++;
	}
	
	return letterCounts;
}

function startLetterCountFunctional(names)
{                                                                  
	return under.countBy(names, (name) => name[0]);  
}
	
module.exports = {averageLettersImperative, averageLettersFunctional, startLetterCountImperative, startLetterCountFunctional};

File: Rust/src/name_util.rs
use std::collections::HashMap;

extern crate itertools;

use itertools::Itertools;

pub fn average_letter_imperative(names: Vec<String>) -> f32 {
    return if names.len() == 0 {0.0} else {
        let mut total = 0.0;
        for name in &names {
            total += name.len() as f32;
        }
        let average = total as f32 / names.len() as f32;
        (average * 100.0).round() /100.0
    }
}

pub fn average_letter_functional(names: &Vec<String>) -> f32 {
    return if names.len() == 0 {0.0} else {
        let total = names.into_iter()
                          .map(|name| name.len())
                          .fold(0,|acc, length| acc+length);
        let average = total as f32 / names.len() as f32;
        (average * 100.0).round() /100.0
    }
}

pub fn start_letter_count_imperative(names: &Vec<String>) -> HashMap<String, i32> {
    let mut dict = HashMap::new();

    for name in names {
        let count = dict.entry(name[0..1].to_string()).or_insert(0);
        *count += 1;
    }
    return dict;
}

pub fn start_letter_count_functional(names:&Vec<String>) -> HashMap<String, i32> {
    names.iter()
         .group_by(|name| name[0..1].to_string())
         .into_iter()
         .map(|(name, group)| (name, group.count() as i32))
         .collect()
}
